### PR TITLE
feat: [IDP-469]: Multiple replica support for events in IDP service

### PR DESCRIPTION
**Pull request by devesh-harness**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/48660)

### CODE REVIEW

**Comment by devesh-harness** [comment]
trigger uts

File: idp-service/src/main/java/io/harness/idp/events/eventlisteners/eventhandler/EntityCrudStreamListener.java [hunk 1]
```diff
@@ -58,20 +62,65 @@ public boolean handleMessage(Message message) {
...
+                Thread.sleep(1000);
+                continue;
+              }
+
+              log.info(
+                  "Lock acquired for processing crud event for resource of type {} with identifier {} for account {} with message id - {}",
+                  entityType, resourceIdentifier, accountIdentifier, messageId);
+
+              try {
+                if (eventMessageHandler != null && entityChangeDTO != null) {
```
Comment (vikyathharekal) [inline idp-service/src/main/java/io/harness/idp/events/eventlisteners/eventhandler/EntityCrudStreamListener.java:96]: Can we simplify this?
eventMessageHandler != null check is already done above. Do we still need this?
entityChangeDTO != null is done twice?

File: idp-service/src/main/java/io/harness/idp/events/eventlisteners/eventhandler/EntityCrudStreamListener.java [hunk 2]
```diff
@@ -58,20 +62,65 @@ public boolean handleMessage(Message message) {
...
+                  if (entityChangeDTO != null) {
+                    eventMessageHandler.handleMessage(message, entityChangeDTO, action);
+                    log.info("Completed processing the crud event with the id {}", messageId);
+                  }
+                }
+              } catch (Exception e) {
+                log.error(
+                    "Error in handeling the crud event with the id {} for entity type {}", messageId, entityType, e);
```
Comment (vikyathharekal) [inline idp-service/src/main/java/io/harness/idp/events/eventlisteners/eventhandler/EntityCrudStreamListener.java:104]: typo `handling`

File: idp-service/src/main/java/io/harness/idp/events/eventlisteners/eventhandler/EntityCrudStreamListener.java [hunk 3]
```diff
@@ -58,20 +62,65 @@ public boolean handleMessage(Message message) {
...
+              } finally {
+                redisLocker.destroy(lock);
+                Thread.sleep(200);
+                log.info(
```
[the remark this answers is not in the document]
Comment (devesh-harness) [reply idp-service/src/main/java/io/harness/idp/events/eventlisteners/eventhandler/EntityCrudStreamListener.java:104]: done, added debug now as multiple logs were showing while i was testing it, so we can add it to debug.

File: idp-service/src/main/java/io/harness/idp/events/eventlisteners/eventhandler/EntityCrudStreamListener.java (hunk 2 above)
Comment (devesh-harness) [reply idp-service/src/main/java/io/harness/idp/events/eventlisteners/eventhandler/EntityCrudStreamListener.java]: done

File: idp-service/src/main/java/io/harness/idp/events/eventlisteners/eventhandler/EntityCrudStreamListener.java (hunk 1 above)
Comment (devesh-harness) [reply idp-service/src/main/java/io/harness/idp/events/eventlisteners/eventhandler/EntityCrudStreamListener.java]: done yes due to multiple refactoring this got missed its not needed.

File: idp-service/src/main/java/io/harness/idp/events/eventlisteners/eventhandler/EntityCrudStreamListener.java [hunk 4]
```diff
@@ -58,20 +62,61 @@ public boolean handleMessage(Message message) {
...
               String.format("Exception in unpacking EntityChangeDTO for id %s", messageId), e);
         }
+
         if (entityChangeDTO != null) {
-          eventMessageHandler.handleMessage(message, entityChangeDTO, action);
-          log.info("Completed processing the crud event with the id {}", messageId);
+          String accountIdentifier = entityChangeDTO.getAccountIdentifier().getValue();
+          String resourceIdentifier = entityChangeDTO.getIdentifier().getValue();
+
+          try {
```
Comment (vigneswara-propelo) [inline idp-service/src/main/java/io/harness/idp/events/eventlisteners/eventhandler/EntityCrudStreamListener.java:78]: do we need this try-catch block? tryToAcquireLock method doesn't throw exception.
Can we also try to move multiple try-catch blocks to separate methods for readability purpose?

Comment (devesh-harness) [reply idp-service/src/main/java/io/harness/idp/events/eventlisteners/eventhandler/EntityCrudStreamListener.java]: yes that was not needed removed it, now only we have tow try catch blocks that should be fine.

**Review by vigneswara-propelo — APPROVED** [review]

**Comment by devesh-harness** [comment]
trigger uts

**Review by vigneswara-propelo — APPROVED** [review]

**Review by vikyathharekal — APPROVED** [review]

**Comment by vikyathharekal** [comment]
trigger review

**Pull request merged**